File: 0221-maximal-square/maximal-square.py
class Solution(object):
    def maximalSquare(self, matrix):
        
        if not matrix:
            return 0
        
        m, n = len(matrix), len(matrix[0])
        res = 0
        
        dp = [[0] * (n+1) for _ in range(m+1)]
        
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                if matrix[i][j] == "1":
                    dp[i][j] = 1 + min(dp[i + 1][j], dp[i][j + 1], dp[i + 1][j + 1])
                    res = max(res, dp[i][j])
        return res * res

File: 0221-maximal-square/maximal-square.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        if(matrix.length==0) return 0;
        int res=0;
        int m=matrix.length; int n= matrix[0].length;
        int [][]dp=new int[m+1][n+1];
        
        for(int i=m-1;i>=0;i--){
            for(int j=n-1;j>=0;j--){
                if(matrix[i][j]=='1'){
                    dp[i][j]=1+Math.min(Math.min(dp[i+1][j],dp[i][j+1]),dp[i+1][j+1]);
                    res=Math.max(res,dp[i][j]);
                }
            }
        }
        return res*res;   
    }
};